File: KDiffPair.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Y
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach

import java.util.HashMap;
import java.util.Map;

public class KDiffPair {
    public int findPairs(int[] nums, int k) {
        Map<Integer,Integer> map = new HashMap<>();
        int count = 0;
         for(int n : nums)
         {
             if(map.containsKey(n))
             {
                 map.put(n, map.get(n) + 1);
             }
             else
                map.put(n, 1);
         }
         for(Map.Entry<Integer,Integer> m : map.entrySet())
         {

             int key = m.getKey();
             int value = m.getValue();

             if(k==0)
             {
                 if(value > 1)
                 {
                     count++;
                 }
             }
             else
             {
                 if(map.containsKey(key+k))
                 {
                     count++;
                 }
             }
         }

         return count;
        }
        public static void main(String[] args)
        {
            int [] nums = {3,1,4,1,5};
            KDiffPair k = new KDiffPair();
            System.out.println(k.findPairs(nums,2));
        }
}

File: kdiffpair.go
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :

// Your code here along with comments explaining your approach
/*
set count = 0
create map of elements and their counts
if length <2 return count
loop through elements and set m[element]++
2 cases if k==0 and element count 1 increment count
after loop if k == 0 return count else
loop through elements again if element + k value element exists and elements count !=-1 increment count and set count = -1
*/
package main

import "fmt"

func findPairs(nums []int, k int) int {
	if len(nums) < 2 {
		return 0
	}
	c := 0
	m := make(map[int]int)

	for i:=0;i<len(nums);i++ {
		v, e := m[nums[i]]
		m[nums[i]] = v + 1
		if e && v == 1 && k == 0 {
			c++
		}
	}
	if k == 0 {
		return c
	}
	for i:=0;i<len(nums);i++ {
		vorg := m[nums[i]]
		_, e := m[nums[i] + k]
		if e {
			if vorg != -1 {
				c++
				m[nums[i]] = -1
			}
		}
	}
	return c
}

func MainKDiffPair() {
	fmt.Println(findPairs([]int {1,2,4,4,3,3,0,9,2,3}, 3)) //expected 2
}
